File: C/Avogadro.cpp
#include<cstdio>
#include<set>

// Initialise all the variables
bool deleted[100005]; // Keep track of all the deleted columns
int arr [3][100005]; // Is the input table
int instance[3][100005]; // Stores #instances in each row
std::set<int> vecset[100005]; // Stores col number of each number
int toDelete[1000005]; // Store cols we want to delete in a ghettoStack
int stackHead = 0;
int size = 0; // output

void push(int n){
    toDelete[stackHead++] = n;
}

int pop(){
    return toDelete[--stackHead];
}


int main(){
    int n, i, j, ph;
    scanf("%d", &n);

    // Scan in the input
    for (i=0; i<3; ++i){
       for (j=0; j<n; ++j){
           scanf("%d", &ph);
           arr[i][j] = ph;
           vecset[ph].insert(j);
           ++instance[i][ph];
        }
        // Push the numbers which don't appear in any row.
        for (j=1; j<=n; ++j){
            if (instance[i][j] == 0) push(j);
        }
    }

    // while we have numbers to delete, delete numbers
    while (stackHead > 0){
        // For each column Q which N appears in:
        int N = pop();
        for (auto Q : vecset[N]){
            // If not already in output, add them in
            if (!deleted[Q]){
                ++size;
                deleted[Q] = true;
                // For the numbers M in the three rows of column Q:
                // reduce their instances by 1 and check if 0
                for (i=0; i<3; ++i){
                    int M = arr[i][Q];
                    --instance[i][M];
                    if (instance[i][M] == 0) push(M);
                }
            }
        }
    }

    printf("%d\n", size);
    return 0;
}

File: Java/Avogadro.java
import java.util.*;
import java.io.*;

public class Avogadro{
    static boolean[] deleted = new boolean[100005]; // Deleted columns
    static int[][] arr = new int[3][100005]; // Input table
    static int[][] instance = new int[3][100005]; // Stores #instances in each row
    static ArrayList<HashSet<Integer>> vecset = new ArrayList<>();
    static Set<Integer> output = new HashSet<>(); // Stores cols we deleted
    static Set<Integer> toDelete = new HashSet<>(); // Stores cols we want to delete
    public static void main(String[] arg){
        try{
            // Scan in input using Buffered Reader
            BufferedReader r = new BufferedReader(new InputStreamReader(System.in));
            int n = Integer.parseInt(r.readLine());

            // Set up the vecset thing, stores col number of each number
            // Set up the deleted columns thing
            for (int i=0; i<100005; i++){
                vecset.add(new HashSet<Integer>());
                deleted[i] = false;
            }

            // For each entry, store in arr, and track the cols it appears in
            // Also take note of the number of instances of each number
            for (int i=0; i<3; ++i){
                String[] ph = r.readLine().split(" ");
                for (int j=0; j<n; ++j){
                    int k = Integer.parseInt(ph[j]);
                    arr[i][j] = k;
                    vecset.get(k).add(j);
                }
            }

            // Keep track of #instances of each number left in each row
            for (int i=0; i<3; i++){
                for (int j=0; j<n; j++) instance[i][arr[i][j]]++;
            }

            // Create a set toDelete which will contain numbers to delete
            for (int i=0; i<3; i++){
                for (int j=0; j<=n; j++){
                    if (instance[i][j] == 0) toDelete.add(j);
                }
            }

            // while toDelete has numbers in it, delete numbers
            while (toDelete.size() > 0){
            // For each column Q which N appears in:
                int N = toDelete.iterator().next();
                toDelete.remove(N);
                for (int Q : vecset.get(N)){
                    // if not already in output, add them in
                    if (!deleted[Q]){
                        output.add(Q);
                        deleted[Q] = true;
                        // for the numbers M in the Q three rows:
                        // reduce their #instances by 1 and check if it's 0
                        for (int i=0; i<3; ++i){
                            int M = arr[i][Q];
                            --instance[i][M];
                            if (instance[i][M] == 0) toDelete.add(M);
                        }
                    }
                }
            }

            System.out.printf("%d\n", output.size());
        }
        catch (IOException e){}
    }
}
